File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        // recursive relation: 
        // rob(i) = max(rob(i-2) + currentHouseValue, rob(i-1))

        if (nums.size() == 0) {
            return 0;
        }

        int prev1 = 0;
        int prev2 = 0;
        // [prev2, prev1, n , n+1, ...]

        for (int num : nums) {
            int temp = max(prev2 + num, prev1);
            prev2 = prev1;
            prev1 = temp;
        }
        return prev1;
    }
};

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # rob every alt house starting from i = 0 or i = 1
        n = len(nums)

        if n == 1:
            return nums[0]
        
        prev, curr = nums[0], max(nums[0], nums[1])
        for i in range(2, n):
            temp = curr
            curr = max(curr, prev + nums[i])
            prev = temp
        
        return curr